### PR TITLE
[RFC] cmake: Add function for adding snippets to the linker file

### PR DESCRIPTION
**Pull request by oyvindronningstad**
add_linker_snippet() allows snippets to be placed in different locations:
 - The noinit, rwdata and rodata output sections
 - Two different locations for placing custom output sections, one location for RAM and another for all other sections.

This PR removes the custom-\*.ld and soc-\*.ld mechanisms which are similar, but more cumbersome to use.

### CODE REVIEW

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -689,6 +689,48 @@ function(zephyr_check_compiler_flag lang option check)
   endif()
 endfunction()
 
+function(add_linker_snippet location snippet)
+  #set(snippet_base "${PROJECT_BINARY_DIR}/include/generated/snippets-")
```
Comment (hakonfam) [inline cmake/extensions.cmake]: remove

File: include/arch/arc/v2/linker.ld [hunk 2]
```diff
@@ -114,6 +114,9 @@ SECTIONS {
 		*(".rodata.*")
 		*(.gnu.linkonce.r.*)
 
+/* Located in generated directory. This file is populated by the
+ * add_linker_snippet() Cmake function. */
```
Comment (hakonfam) [inline include/arch/arc/v2/linker.ld]: I think you have to have the ending `*/` on a new line as discussed earlier.

File: cmake/extensions.cmake (hunk 1 above)
Comment (oyvindronningstad) [reply cmake/extensions.cmake]: Done

File: include/arch/arc/v2/linker.ld (hunk 2 above)
Comment (oyvindronningstad) [reply include/arch/arc/v2/linker.ld]: Done

**Comment by SebastianBoe** [comment]
wow, nice check

**Comment by carlescufi** [comment]
@oyvindronningstad any reason this cannot go upstream?

**Comment by oyvindronningstad** [comment]
> @oyvindronningstad any reason this cannot go upstream?

No, I wanted to RFC it here first, but I'm hoping it can go upstream. I'll open a PR.

File: arch/arm/core/cortex_m/CMakeLists.txt [hunk 3]
```diff
@@ -9,3 +9,12 @@ zephyr_library_sources(
...
   exc_manage.c
   )
+
+add_linker_snippet_ifdef(CONFIG_SW_VECTOR_RELAY RAM_SECTIONS
+  "/* Reserved 4 bytes to save vector table base address */
+  SECTION_PROLOGUE(.vt_pointer,(NOLOAD),)
+  {
+          *(.vt_pointer_section)
+          *(\".vt_pointer_section.*\")
+  }"
```
Comment (SebastianBoe) [inline arch/arm/core/cortex_m/CMakeLists.txt]: Thoughts on having the LD script code in .ld files instead?

Benefits as I see it;

- Easy to find LD script code, (look for .ld files).
- Easier for an editor to help with formatting and syntax highlighting.
- User does not need to know which symbols need to be escaped.
- Consistent with how we add other files, e.g. c files.

Comment (oyvindronningstad) [reply arch/arm/core/cortex_m/CMakeLists.txt]: Sure, I see no big reasons not to. So the Cmake would look something like this maybe?

    zephyr_linker_sources(<location> <files>)
e.g:

    zephyr_linker_sources(RODATA
      ${CMAKE_CURRENT_LIST_DIR}/mylinkerscript.ld
    )

**Comment by oyvindronningstad** [comment]
PR open upstream

**Comment by hakonfam** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/15112

File: Kconfig.zephyr [hunk 4]
```diff
@@ -143,24 +143,6 @@ config CUSTOM_LINKER_SCRIPT
...
-	help
-	  Include a customized linker script fragment for inserting additional
-	  data and linker directives into the data section.
-
-config CUSTOM_SECTIONS_LD
-	bool "Include custom-sections.ld"
-	help
-	  Include a customized linker script fragment for inserting additional
-	  arbitrary sections.
-
```
Comment (SebastianBoe) [inline Kconfig.zephyr]: Before dropping support for CUSTOM_RODATA_LD we might want to first mark it as deprecated to give users a reasonable window of time to migrate.

File: arch/arm/core/cortex_m/CMakeLists.txt (hunk 3 above)
Comment (SebastianBoe) [reply arch/arm/core/cortex_m/CMakeLists.txt]: yes

Comment (SebastianBoe) [reply arch/arm/core/cortex_m/CMakeLists.txt]: check out how zephyr_sources is implemented to see how to get rid of the CMAKE_CURRENT_LIST_DIR.

**Comment by SebastianBoe** [comment]
Closing as this is now an upstream RFC.

File: arch/arm/core/cortex_m/CMakeLists.txt (hunk 3 above)
Comment (oyvindronningstad) [reply arch/arm/core/cortex_m/CMakeLists.txt]: Fixed in upstream PR

**Pull request closed without merging**